File: src/game/bounties/rotation.ts
import { BOUNTIES_PER_HOUR, BOUNTY_CATALOG } from './catalog'
import type { BountyDefinition, HourlyBountyBoard } from './types'

/** UTC hour bucket, e.g. 2026-08-12T13 */
export function bountyHourKey(nowMs: number = Date.now()): string {
  const date = new Date(nowMs)
  const y = date.getUTCFullYear()
  const m = String(date.getUTCMonth() + 1).padStart(2, '0')
  const d = String(date.getUTCDate()).padStart(2, '0')
  const h = String(date.getUTCHours()).padStart(2, '0')
  return `${y}-${m}-${d}T${h}`
}

export function bountyHourExpiresAtMs(hourKey: string, nowMs: number = Date.now()): number {
  const parsed = Date.parse(`${hourKey}:00:00.000Z`)
  if (!Number.isFinite(parsed)) return nowMs + 3_600_000
  return parsed + 3_600_000
}

function hashString(input: string): number {
  let hash = 2166136261
  for (let i = 0; i < input.length; i += 1) {
    hash ^= input.charCodeAt(i)
    hash = Math.imul(hash, 16777619)
  }
  return hash >>> 0
}

/** Deterministic hourly sample from the bounty catalog. */
export function hourlyBountyBoard(nowMs: number = Date.now()): HourlyBountyBoard {
  const hourKey = bountyHourKey(nowMs)
  const pool = [...BOUNTY_CATALOG]
  const selected: BountyDefinition[] = []
  let seed = hashString(hourKey)
  while (selected.length < BOUNTIES_PER_HOUR && pool.length > 0) {
    seed = (Math.imul(seed, 1664525) + 1013904223) >>> 0
    const index = seed % pool.length
    selected.push(pool.splice(index, 1)[0]!)
  }
  return {
    hourKey,
    expiresAtMs: bountyHourExpiresAtMs(hourKey),
    bounties: selected,
  }
}

File: packages/ik_rules/lib/src/bounties/rotation.dart
import '../js_compat.dart';
import '../time.dart';
import 'catalog.dart';
import 'types.dart';

const int _uint32Mask = 0xFFFFFFFF;
const num _hourMs = 3600000;

/// UTC hour bucket, e.g. 2026-08-12T13
String bountyHourKey(num nowMs) {
  final date = DateTime.fromMillisecondsSinceEpoch(nowMs.toInt(), isUtc: true);
  final month = date.month.toString().padLeft(2, '0');
  final day = date.day.toString().padLeft(2, '0');
  final hour = date.hour.toString().padLeft(2, '0');
  return '${date.year}-$month-${day}T$hour';
}

num bountyHourExpiresAtMs(String hourKey, num fallbackNowMs) {
  final parsed = jsDateParse('$hourKey:00:00.000Z');
  if (!parsed.isFinite) return fallbackNowMs + _hourMs;
  return parsed + _hourMs;
}

int _hashString(String input) {
  var hash = 2166136261;
  for (var i = 0; i < input.length; i += 1) {
    hash ^= input.codeUnitAt(i);
    hash = jsImul(hash, 16777619);
  }
  return hash & _uint32Mask;
}

/// Deterministic hourly sample from the bounty catalog.
HourlyBountyBoard hourlyBountyBoard(num nowMs) {
  final hourKey = bountyHourKey(nowMs);
  final pool = [...bountyCatalog];
  final selected = <BountyDefinition>[];
  var seed = _hashString(hourKey);
  while (selected.length < bountiesPerHour && pool.isNotEmpty) {
    seed = (jsImul(seed, 1664525) + 1013904223) & _uint32Mask;
    selected.add(pool.removeAt(seed % pool.length));
  }
  return HourlyBountyBoard(
    hourKey: hourKey,
    expiresAtMs: bountyHourExpiresAtMs(hourKey, nowMs),
    bounties: selected,
  );
}
